### DISCUSSION TITLE
Charts visualization on Streamlit web app

### DISCUSSION THREAD

**Discussion by marcello-calabrese** [sinaptik-ai/pandas-ai#221] [Q&A]
Hi, is there a better way to visualize charts through pandasai on Streamlit? I will explain, so to show the charts on streamlit when I ask to plot, I used matplotlib.use('TkAgg'), it works that basically show the chart in a new window but the request keeps running.

It might be a better way to render matplotlib through pandasai on a streamlit. Hope so.

Thanks a lot, 

Marcello

**Comment by gventuri** [comment]
Hi @marcello-calabrese! Might be necessary to instruct the model to use `block=False` when plotting the chart. This should fix the issue hopefully!
I'll look into it further to figure it out, thanks a lot for reporting.

P.s. if you have any alternative approach in mind, feel free to share!

**Reply by marcello-calabrese** [reply]
Ciao @gventuri (immagino sei italiano 😊), thanks for your quick response. I will try to use `block=false` otherwise not sure if I create an input_text query extra for charts, and use st.py plot, will try

**Comment by gventuri** [comment]
Ciao @marcello-calabrese (yes, I'm from Italy 😊)! You should actually instruct the model to use `block=false` as it plots the chart for the time being, so try to mention it in the prompt. We will soon work on a patch to have it done automatically!

Let me know if it works!

**Reply by marcello-calabrese** [reply]
Grazie @gventuri , I will try and let you know.

**Reply by marcello-calabrese** [reply]
Hi again @gventuri , I tried to add `block=False`, first in `pandas_ai = PandasAI(llm=llm, conversational=True, verbose=False, block=False)` and then I tried it in `st.write(pandas_ai.run(df,question, is_conversational_answer=True, block=False))` , and it didnt work. It says `block=False` not valid. Am I doing something wrong? st.write is the streamlit code to show the '' answer'' of the chatbot. The message error is: `TypeError: PandasAI.run() got an unexpected keyword argument 'block'`
Happy to provide more details if needed. Thanks :-)

**Reply by gventuri** [reply]
@marcello-calabrese sorry I was mentioning replacing `plt.show()` with `plt.show(block=False)` from the output.
However I guess the point here is that streamlit expects `st.pyplot()` instead of `plt.show()`, we should find a way to parse the output and replacing it before running the code.

I'll open an issue and try to work on it asap!

**Reply by gventuri** [reply]
@marcello-calabrese check it out: https://pandas-ai.readthedocs.io/en/latest/middlewares/

We now support middlewares to address cases as the one you mentioned. By passing the streamlit middleware it should now work. Let me know :)

**Reply by marcello-calabrese** [reply]
Thanks @gventuri I will try and find out. I will let you know

**Reply by marcello-calabrese** [reply]
CIao @gventuri , I tried with middleware following the guide you mentioned above, of course after updating pandasai in the environment, but it gives me the error saying streamlitmiddleware does not exist. Let me know if you would like me to add the error screenshot. Thanks,

**Reply by gventuri** [reply]
@marcello-calabrese don't forget to import it:

```
from pandasai import PandasAI
from pandasai.middlewares.streamlit import StreamlitMiddleware

pandas_ai = PandasAI(llm, middlewares=[StreamlitMiddleware()])
```

Let me know if it works for you!

**Reply by marcello-calabrese** [reply]
Hi @gventuri , thanks i imported the Streanlit Middleware as you suggested above, but it doesnt create any chart, it only respond as below (see the screenshot);

![image](https://github.com/gventuri/pandas-ai/assets/74682725/57d1567f-4d69-467a-911f-81e1f0e16825)

**Reply by gventuri** [reply]
@marcello-calabrese thanks a lot for sharing. Would you mind sharing the code you are using so that I can try to reproduce it?

**Reply by gventuri** [reply]
@marcello-calabrese just released v0.5.5 that should fix this error!

**Comment by marcello-calabrese** [comment]
> @marcello-calabrese thanks a lot for sharing. Would you mind sharing the code you are using so that I can try to reproduce it?

Hi, see the code below thanks. the code below is not with the version updated to 0.5.5, will try soon, thanks

# instantiate LLM with PandasAI

llm = OpenAI(api_token='sk-XXXXXXXXXXXXXXXXXXXXXXXXXXXXXX', engine='gpt-3.5-turbo')

pandas_ai = PandasAI(llm, middlewares=[StreamlitMiddleware()], verbose=False)

if st.button('Generate Response'):
        if question:
            with st.spinner('Generating response, please wait...'):
                # get the response from the chatbot
                #pandas_ai.run(df,question, is_conversational_answer=True)
                # show the response
                st.write(pandas_ai.run(df,question, is_conversational_answer=True))

**Reply by gventuri** [reply]
I expect the code to work with 0.5.5, as I'm successfully running something very similar! Feel free to give it a try and let me know :)

**Reply by laodev1** [reply]
Just wanted to confirm that this does work in > 0.5.5

**Reply by marcello-calabrese** [reply]
Hi @gventuri , sorry for my late reply, I job took all of my time these days :-( and dint let me do what I love: data analytics and data science.

I can confirm that with the 0.5.5 version it works perfectly Thanks a lot GRAZIE :-)

**Reply by gventuri** [reply]
@laodev1 @marcello-calabrese thanks a lot for letting me know 🚀
